src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt

```
@@ -227,9 +245,41 @@ class UserService(
         val passwordHash = PasswordUtils.hashPassword(request.password)
         val user = userRepo.createUser(request, passwordHash)
 
-        // Generate JWT tokens
-        val (accessToken, refreshToken) = jwtService.generateTokens(user.id)
-        GrpcContext.setAuthCookiesInContext(accessToken, refreshToken)
+        // Generate and save verification token for the user
+        val verificationToken = NanoId.generate(VERIFICATION_TOKEN_LENGTH)
+        verificationTokenRepo.saveVerificationToken(user.id, verificationToken)
+
+        // Send verification email
+        val verificationLink = "${envReader.env.miscellaneous.frontendBaseUrl}/verifyemail?token=$verificationToken"
+        emailService.sendEmailVerificationEmail(
+            user.email,
+            EmailData.EmailVerification(
+                user.firstName,
+                user.lastName,
+                verificationLink,
+            ),
+        )
+
+        return Base.Nothing.getDefaultInstance()
+    }
+
+    override suspend fun verifyEmail(request: Authentication.VerifyEmailRequest): Base.Nothing {
+        val verificationToken = verificationTokenRepo.getVerificationTokenByValue(request.token)
+            ?: throw SnowballRException.VerificationTokenNotFoundException(request.token)
+
+        // Check whether the user exists
+        val user = userRepo.getUserById(verificationToken.userId)
+
+        // Update the user's status to active
+        val updatedUser = user.copy(status = UserStatus.USER_STATUS_ACTIVE)
+        val userUpdate = GrpcUser.Update.newBuilder()
+            .setUser(updatedUser.toGrpcUser())
+            .setMask(FieldMask.newBuilder().addPaths("user.status").build())
+            .build()
+        userRepo.updateUser(userUpdate)
```

AFAIK, there is no use-case for a user modifying a user status.
Using `userRepo.updateUser` requires the update logic of users to allow changing the user status, though. I think this should be a custom (internal) `userRepo` call instead.

---

I'm not completely sure what you mean. How would the user be table to change the status?
Yes, this call allows this, but there is no way the user is actually able to do this from the frontend, or is he?

I talked with @Slartibartfass2 about this at the time and we agreed to simply add it here. Or is there really a use case where the user could set his own status? How would he be able to do this?

Even then, wouldn't it make more sense to keep one `updateUser` function in the repo layer but rather validate in the `updateUser` call in the service layer that only valid properties can be updated? Then we could simply throw an exception if there is anything that is not allowed.

---

The idea was that the repo method doesn't contain any business logic and therefore it should just update whatever is requested.

There's an issue that the paths are checked in the facade layer so that the requests only contain allowed paths. This would prevent anyone from updating protected fields.

---

This just looked very much like a user could send a similar request. If there is additional validation already in place, then it is totally fine like this.

Just feels a bit weird constructing what is essentially a user request internally.